### PR TITLE
Add align_ptr.h

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
This functionality will be useful for future optimizations.

**Comment by gperciva** [comment]
(The next time we run `make Makefiles`, it will get added as a dependency for `crypto_aes_aesni.c`.  I didn't bother including those changes in this PR, because they'll get picked up the next time we run it.)

File: util/align_ptr.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#define _ALIGN_PTR_H_
+
+/**
+ * ALIGN_PTR_DECL(type, name, num):
+ * Declare a variable array called ${name}, large enough to contain an array
+ * of ${num} values of ${type}.  Must not be used until ALIGN_PTR_INIT() has
+ * been called.  This macro may also create an additional variable called
+ * "${name}_buf".
+ */
+#define ALIGN_PTR_DECL(type, name, num)					\
```
Comment (gperciva) [inline util/align_ptr.h:11]: Hmm, I just realized that this won't help in `crypto_aesctr.c`.  In that case, we want a `uint8_t` array that's aligned to satisfy `__m128i`.

I'll mess around a bit more, and probably add a REBASE commit to this PR.

File: util/align_ptr.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#ifndef _ALIGN_PTR_H_
+#define _ALIGN_PTR_H_
+
```
Comment (gperciva) [inline util/align_ptr.h:3]: whoops, this should include `<stdint.h>`.

**Comment by gperciva** [comment]
Aha, I forgot to add it to `tests/buildsingles/Makefile-headers`.  That shows a few problems, will be fixed soon.

**Comment by gperciva** [comment]
Ok, ready for review again.

**Comment by cperciva** [comment]
"aesci" in commit message

File: util/align_ptr.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+#define ALIGN_PTR_DECL(type, name, num, alignsize)			\
+	uint8_t name##_buf[num * sizeof(type) + (alignsize - 1)];	\
+	type * name
+
+/**
+ * ALIGN_PTR_INIT(name, alignsize):
+ * Initialize the variable called ${name} to point to an array which is
+ * aligned to ${alignsize} bytes.  They must have previously been declared
+ * with ALIGN_PTR_DECL(${name}, ${alignsize}).
```
Comment (cperciva) [inline util/align_ptr.h:23]: This deserves a comment "Requires that casting a pointer to uintptr_t behaves naturally, i.e. yields a byte offset relative to aligned memory."

The semantics required for uintptr_t are far weaker -- the values don't have to be meaningful at all, beyond returning the same pointer if you cast them back -- but every platform I'm aware of provides this (I think even including CHERI, which breaks all sorts of pointer-related assumptions).

**Comment by gperciva** [comment]
Revised PR, including two comments about uintptr_t assumptions in align_ptr.h, and adding a slightly edited version of your comment about "the actual semantics required for uintpt_t" to the commit message.

**Pull request merged**